BUILD.md

```diff
@@ -100,16 +110,18 @@ architecture:
 conan profile update settings.arch=x86_64 default
 ```
 
-If you have multiple compilers installed on your platform,
+**If you have multiple compilers** installed on your platform,
 then you'll need to make sure that Conan and CMake select the one you want to
 use.
-This setting will set the correct variables (`CMAKE_<LANG>_COMPILER`) in the
-generated CMake toolchain file:
-
+Update the `conf.tools.build:compiler_executables` setting in order to set the correct variables (`CMAKE_<LANG>_COMPILER`) in the
+generated CMake toolchain file.
+For example, on Ubuntu 20, you may have gcc at `/usr/bin/gcc` and g++ at `/usr/bin/g++`; if that is the case, you can select those compilers with:
 ```
-conan profile update 'conf.tools.build:compiler_executables={"c": "<path>", "cpp": "<path>"}' default
+conan profile update 'conf.tools.build:compiler_executables={"c": "/usr/bin/gcc", "cpp": "/usr/bin/g++"}' default
```

Comment:
This is unnecessary because `cc` and `c++` are the defaults that CMake will use (and CMake is the build system used by almost all of our dependencies), and they will resolve to `/usr/bin/gcc` and `/usr/bin/g++` on the default `PATH`. If builders require this level of instruction, then they likely don't understand that `/usr/bin/g++` is the wrong choice. It is more likely to be the default installed compiler and not the newer version they presumably installed to build rippled. This example seems like it's just going to confuse people more. They will fail to build with the defaults, then set these values that match the defaults, it won't fix their problem, and they'll think this instruction is wrong.

What was the misunderstanding that led to this change?

Reply:
> This is unnecessary because `cc` and `c++` are the defaults that CMake will use (and CMake is the build system used by almost all of our dependencies), and they will resolve to `/usr/bin/gcc` and `/usr/bin/g++` on the default `PATH`. If builders require this level of instruction, then they likely don't understand that `/usr/bin/g++` is the wrong choice. It is more likely to be the default installed compiler and not the newer version they presumably installed to build rippled.

How do you know that `/usr/bin/g++` is the wrong choice?

> This example seems like it's just going to confuse people more. They will fail to build with the defaults, then set these values that match the defaults, it won't fix their problem, and they'll think this instruction is wrong.
> 
> What was the misunderstanding that led to this change?

People have failed to build with the instructions as given because they don't provide a path that could ever work:

```
CMake Error at CMakeLists.txt:49 (project):
  The CMAKE_CXX_COMPILER:

    <path>

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.
```

Reply:
It's because `/usr/bin/g++` is the default value (when it exists on a platform). If you have to set these settings, it's because you have multiple compilers installed and the default compiler, the one picked by CMake or b2 by default, is not the one that can build rippled. If you set this setting to the default value, it's the same as not setting it. The only reason to set this setting is to override the default value.

`<path>` is a placeholder. Isn't that obvious? When people see that notation in any other command line, including CMake help strings, they know what it means: it means this is a parameter and you have to substitute the value that makes sense for you. Do we really have readers that don't know this? Genuine question. It starts to feel like this document is held to an impossible standard and prohibited from following common practices in software documentation.

The reason we have to use a placeholder is because we cannot tell what the right value is for any particular user. It might be any of the following on common platforms:

- `/usr/bin/g++-10`
- `/usr/bin/g++-11`
- `/usr/bin/g++-12`
- `/usr/bin/g++-13`
- `/usr/local/bin/g++-10`
- `/usr/local/bin/g++-11`
- `/usr/local/bin/g++-12`
- `/usr/local/bin/g++-13`
- `/opt/local/bin/g++-10`
- `/opt/local/bin/g++-11`
- `/opt/local/bin/g++-12`
- `/opt/local/bin/g++-13`
- `/opt/homebrew/bin/g++`
- `/opt/homebrew/bin/g++-10`
- `/opt/homebrew/bin/g++-11`
- `/opt/homebrew/bin/g++-12`
- `/opt/homebrew/bin/g++-13`
- `/usr/local/bin/clang-14'
- `/usr/local/bin/clang-15'
- `/usr/local/bin/clang-16'

This list is not exhaustive. It could literally be anywhere. Only the builder knows exactly where.

How do we get people to stop blindly copying commands, or to read the error message they get? The error message you pasted makes it immediately clear what the problem is, doesn't it?

These two sets of commands are coupled. They should either both use a placeholder or both use an example path. Though, if readers cannot recognize a placeholder that needs to be replaced, then how are they going to recognize an example that needs to be replaced?

Reply:
> It's because `/usr/bin/g++` is the default value (when it exists on a platform). If you have to set these settings, it's because you have multiple compilers installed and the default compiler, the one picked by CMake or b2 by default, is not the one that can build rippled or its dependencies. If you set this setting to the default value, it's the same as not setting it. The only reason to set this setting is to override the default value.

That was not clear from the existing doc. I had styled **If you have multiple compilers** in bold, which would likely have helped, but c8f17324f6b4aa5d8e2d8f7994fb7b76a7d4bf49 makes it even more clear by giving it a new sub-heading, "Multiple compilers".

> `<path>` is a placeholder. Isn't that obvious? When people see that notation in any other command line, including CMake help strings, they know what it means: it means this is a parameter and you have to substitute the value that makes sense for you. Do we really have readers that don't know this? Genuine question. It starts to feel like this document is held to an impossible standard and prohibited from following common practices in software documentation.

CMake help strings are a different context than `BUILD.md`. In `BUILD.md`, anything you provide in code blocks (```) is expected to be copyable, and in fact, GitHub makes it easy to do this by providing a one-click "copy" button near the right side of the code block. To see this, open https://github.com/XRPLF/rippled/blob/develop/BUILD.md in a web browser.

> The reason we have to use a placeholder is because we cannot tell what the right value is for any particular user. It might be any of the following on common platforms:
> 
> * `/usr/bin/g++-10`
> * `/usr/bin/g++-11`
> * `/usr/bin/g++-12`
> * `/usr/bin/g++-13`
> * `/usr/local/bin/g++-10`
> * `/usr/local/bin/g++-11`
> * `/usr/local/bin/g++-12`
> * `/usr/local/bin/g++-13`
> * `/opt/local/bin/g++-10`
> * `/opt/local/bin/g++-11`
> * `/opt/local/bin/g++-12`
> * `/opt/local/bin/g++-13`
> * `/opt/homebrew/bin/g++`
> * `/opt/homebrew/bin/g++-10`
> * `/opt/homebrew/bin/g++-11`
> * `/opt/homebrew/bin/g++-12`
> * `/opt/homebrew/bin/g++-13`
> * `/usr/bin/clang++-14`
> * `/usr/bin/clang++-15`
> * `/usr/bin/clang++-16`
> * `/usr/local/bin/clang++-14`
> * `/usr/local/bin/clang++-15`
> * `/usr/local/bin/clang++-16`
> * ...
> 
> This list is not exhaustive. It could literally be anywhere. Only the builder knows exactly where.

Of course; that's why the document says "Replace `/usr/bin/gcc` and `/usr/bin/g++` with paths to the desired compilers."

That said, in most of the cases I've seen, `/usr/bin/gcc` and `/usr/bin/g++` are actually the right values. If they're not, the user can easily change them.

Furthermore, when providing an example command that someone can use to change the default, it is helpful to provide the default values in that example. Not only does it make it obvious what they are changing the setting _from_, but it also makes it easy to change the setting back to the default, if that is ever desired.

> How do we get people to stop blindly copying commands, or to read the error message they get?

You can remove the code block formatting (```). As for error messages, some will read them. The better solution is to prevent the error message from occurring in the first place.

> The error message you pasted makes it immediately clear what the problem is, doesn't it?

Yes, but it does not make clear what the solution is. `conan profile update 'conf.tools.build:compiler_executables={"c": "/usr/bin/gcc", "cpp": "/usr/bin/g++"}' default` was the solution for some users.

> These two sets of commands are coupled. They should either both use a placeholder or both use an example path. Though, if readers cannot recognize a placeholder that needs to be replaced, then how are they going to recognize an example that needs to be replaced?

It's not about "cannot" - it's about what they actually do, in practice, when trying to build the project while expending the minimal amount of time possible. Time is precious. We're all going to die. If we can change the instructions to save developers 10 minutes of time, all else equal, then we should do it.

As stated, it's not actually true that the example "needs to be replaced", since `/usr/bin/gcc` and `/usr/bin/g++` actually do work for users. And if you're saying that those are the defaults and the command should never be used if those are the desired compilers, then let's remove the code block so it doesn't look like this is a setting that users generally need to set.

Reply:
notes:
- the default is also platform specific
- most dependencies use CMake; boost uses b2

with CMake: when you don't specify compiler on command line, it will search. on Linux, it should start its search at /usr/bin/cc. on Ubuntu, it points to /etc/alt/cc which points to /usr/bin/gcc

Reply:
We'll run with this and if it becomes a problem, go ahead and open a PR in the future to change it again.